src/bsp/src/posix/src/InterlocManager.cpp

```diff
@@ -0,0 +1,148 @@
+#include "InterlocManager.h"
+#include "BSP.h"
+#include <Task.h>
+#include <bsp/BSPContainer.h>
+#include <gazebo_msgs/ModelStates.h>
+#include <regex>
+#include <tf2/LinearMath/Transform.h>
+#include <tf2_geometry_msgs/tf2_geometry_msgs.h>
+
+#define ROBOT_CENTER_SUFFIX "/base_link"
+#define HIVEBOARD_SUFFIX "/hiveboard"
+
+InterlocManager::InterlocManager(ILogger& logger) :
+    m_logger(logger),
+    m_tfListener(m_tfBuffer),
+    m_positionUpdateCallback(nullptr),
+    m_positionUpdateContext(nullptr) {}
+
+tf2::Transform InterlocManager::computeRelativeTransform(
+    const geometry_msgs::Pose& currentAgentPose,
+    const geometry_msgs::TransformStamped& currentAgentHBTransform,
+    const geometry_msgs::Pose& distantAgentPose,
+    const geometry_msgs::TransformStamped& distantAgentHBTransform) {
+
+    // Convert the poses into the HB frame
+    geometry_msgs::Pose currentAgentPoseHBFrame;
+    geometry_msgs::Pose distantAgentPoseHBFrame;
+
+    tf2::doTransform(currentAgentPose, currentAgentPoseHBFrame, currentAgentHBTransform);
+    tf2::doTransform(distantAgentPose, distantAgentPoseHBFrame, distantAgentHBTransform);
+
+    tf2::Transform currentAgentTf;
+    tf2::Transform distantAgentTf;
+    tf2::fromMsg(currentAgentPoseHBFrame, currentAgentTf);
+    tf2::fromMsg(distantAgentPoseHBFrame, distantAgentTf);
+
+    // Equivalent of currentAgentPose - distantAgentPose
+    return distantAgentTf.inverseTimes(currentAgentTf);
+}
+
+double InterlocManager::getDistance(const tf2::Transform& transform) {
+    return sqrt(pow(transform.getOrigin().x(), 2) + pow(transform.getOrigin().y(), 2) +
+                pow(transform.getOrigin().z(), 2));
+}
+
+double InterlocManager::getOrientation(const tf2::Transform& transform) {
+    double roll;
+    double pitch;
+    double yaw;
+    tf2::Matrix3x3(transform.getRotation()).getRPY(roll, pitch, yaw);
+
+    return yaw;
+}
+
+std::optional<geometry_msgs::TransformStamped> InterlocManager::getHiveBoardTransform(
+    const std::string& agentName) {
+    geometry_msgs::TransformStamped transform;
+
+    try {
+        transform = m_tfBuffer.lookupTransform(agentName + HIVEBOARD_SUFFIX,
+                                               agentName + ROBOT_CENTER_SUFFIX, ros::Time(0));
+    } catch (tf2::TransformException& ex) {
+        m_logger.log(LogLevel::Warn, "Could not find transform to hiveboard for %s",
+                     agentName.c_str());
+        m_logger.log(LogLevel::Warn, "%s", ex.what());
+    }
+
+    return transform;
```

Comment:
So garbage is returned if there is an exception? Why not return {} since it's an optinal?

Reply:
My bad, I forgot to return inside of the catch